### PR TITLE
export and import dump from jsonl files

### PR DESCRIPTION
**Pull request by hdiniz**
Resolves https://github.com/opencollective/opencollective/issues/7491
Resolves https://github.com/opencollective/opencollective/issues/7489

### CODE REVIEW

File: scripts/smart-dump.ts [hunk 1]
```diff
@@ -40,31 +42,30 @@ program.command('dump [recipe] [env]').action(async (recipe, env) => {
   const date = new Date().toISOString().substring(0, 10);
   const hash = md5(JSON.stringify({ entries, defaultDependencies, date })).slice(0, 5);
   const filename = `${date}.${hash}`;
-  let docs = [];
+  const seenModelRecords: Set<string> = new Set();
```
Comment (kewitz) [inline scripts/smart-dump.ts:45]: 🫰

File: scripts/smart-dump.ts [hunk 2]
```diff
@@ -40,31 +42,30 @@ program.command('dump [recipe] [env]').action(async (recipe, env) => {
...
   for (const entry of entries) {
     logger.info(`>>> Traversing DB for entry ${entries.indexOf(entry) + 1}/${entries.length}...`);
-    const newdocs = await traverse({ ...entry, defaultDependencies, parsed });
-    docs.push(...newdocs);
+    await traverse({ ...entry, defaultDependencies, parsed }, ei => {
+      const modelRecordKey = `${ei.model}.${ei.id}`;
+      if (!seenModelRecords.has(modelRecordKey)) {
+        dumpFile.write(JSON.stringify(ei) + os.EOL);
+        seenModelRecords.add(modelRecordKey);
+      }
```
Comment (kewitz) [inline scripts/smart-dump.ts:57]: Awesome.

File: scripts/smart-dump.ts [hunk 3]
```diff
@@ -40,31 +42,30 @@ program.command('dump [recipe] [env]').action(async (recipe, env) => {
...
+  logger.info(`>>> Dumping... to dbdumps/${filename}.jsonl`);
+  const dumpFile = fs.createWriteStream(`dbdumps/${filename}.jsonl`);
   for (const entry of entries) {
     logger.info(`>>> Traversing DB for entry ${entries.indexOf(entry) + 1}/${entries.length}...`);
-    const newdocs = await traverse({ ...entry, defaultDependencies, parsed });
-    docs.push(...newdocs);
+    await traverse({ ...entry, defaultDependencies, parsed }, ei => {
+      const modelRecordKey = `${ei.model}.${ei.id}`;
+      if (!seenModelRecords.has(modelRecordKey)) {
+        dumpFile.write(JSON.stringify(ei) + os.EOL);
```
Comment (kewitz) [inline scripts/smart-dump.ts:55]: Perhaps we should await the write to be done here.

File: server/lib/import-export/export.ts [hunk 4]
```diff
@@ -105,7 +98,10 @@ export const traverse = async ({
     } else {
       records.forEach(r => parsed[model].add(r.id));
     }
-    acc.push(...records);
+
+    records.forEach(element => {
+      callback(element);
```
Comment (kewitz) [inline server/lib/import-export/export.ts:103]: Maybe await for each write to be done?

File: scripts/smart-dump.ts (hunk 3 above)
Comment (hdiniz) [reply scripts/smart-dump.ts:68]: The writes are syncronous

File: scripts/smart-dump.ts [hunk 5]
```diff
@@ -34,37 +38,51 @@ program.command('dump [recipe] [env]').action(async (recipe, env) => {
     recipe = './smart-dump/defaultRecipe.js';
   }
 
+  const tempDumpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'oc-export-'));
```
Comment (Betree) [inline scripts/smart-dump.ts:41]: Not sure why the dir name ends with `-`, you maybe wanted to add a unique ID there?

File: scripts/smart-dump.ts [hunk 6]
```diff
@@ -34,37 +38,51 @@ program.command('dump [recipe] [env]').action(async (recipe, env) => {
...
   // eslint-disable-next-line @typescript-eslint/no-var-requires
   const { entries, defaultDependencies } = require(recipe);
   const parsed = {};
   const date = new Date().toISOString().substring(0, 10);
   const hash = md5(JSON.stringify({ entries, defaultDependencies, date })).slice(0, 5);
-  const filename = `${date}.${hash}`;
-  let docs = [];
+  const seenModelRecords: Set<string> = new Set();
+
+  const gitRevision = execSync('git describe --always --abbrev=0 --match "NOT A TAG" --dirty="*"').toString().trim();
```
Comment (Betree) [inline scripts/smart-dump.ts:51]: That could fail in production, where (I believe) `git` is not available. Not a blocker, but we should ideally implement a fallback.

**Review by Betree — APPROVED** [review]
Awesome!

File: scripts/smart-dump.ts (hunk 5 above)
Comment (hdiniz) [reply scripts/smart-dump.ts:41]: `mkdtempSync` generates a unique temp directory with a given prefix, so this results in something like `/tmp/os-export-c122Fuytasd`

**Pull request merged**